### DISCUSSION TITLE
Relay Pagination Plugin does not work if relay-spec compliant connection object is not returned directly as the result of a query.

### DISCUSSION THREAD

**Discussion by xoldd** [urql-graphql/urql#2598] [Q&A] [answered]
Let's say we have a query field called `posts` to fetch list of posts. I've implemented it in a way so that we get both `clientErrors` and `connection` object in a returned object of type `PostsConnectionPayload`.  `clientErrors` here is a list of errors which the client could make while making the query(like not providing any value for 'first` or 'last` or providing `first` with `before` etc.) and `connection` is the relay-spec compliant connection object. Urql's relay plugin doesn't seem to work with this approach. In the cache exchange resolvers field I did add the `relayPagination()` thingy to my custom `PostsConnectionQueryPayload` -> `connection` field. It didn't work and I'm getting many errors in the browser console. I tested another connection field which directly returned the connection object on the root `Query` field and it works perfectly fine.

![Screenshot from 2022-08-11 22-34-17](https://user-images.githubusercontent.com/82873939/184193200-75b45be9-143a-4f10-83e1-14b81ca424d6.png)
![Screenshot from 2022-08-11 22-35-01](https://user-images.githubusercontent.com/82873939/184193205-b36d3579-7ab9-46bc-964f-e52da68bb275.png)
![Screenshot from 2022-08-11 22-39-56](https://user-images.githubusercontent.com/82873939/184193224-eb5e99e5-3f55-41a1-a22b-38e1c577de0d.png)
![Screenshot from 2022-08-11 22-40-07](https://user-images.githubusercontent.com/82873939/184193237-e5bc5073-d5cd-4edc-86c1-9fccf7fdf281.png)

In the images you can see that the plugin works for query field user because I'm directly returning the connection object. But it doesn't work for posts query field and instead the throws many errors in the console. Any fix for this?

**Comment by kitten** [comment, marked as answer]
I'm sorry, I don't really understand what you're trying to describe, since unfortunately all examples here you posted are of working queries and not broken ones.

However, am I correct in assuming you mean a field like `A.field(first: 10, after: $after)` which then returns a type that has a `connection` field which then leads to the real connection?

If that's the case that's not actually compliant with the Relay pagination spec: https://relay.dev/graphql/connections.htm#sec-Arguments

As stated there:

> A field that returns a Connection Type must include forward pagination arguments, backward pagination arguments, or both.

The connection type must be returned directly by the field that accepts the pagination arguments.
You mentioned something like errors, so I assume you've decided that you want to cache errors and those special errors are different than GraphQLErrors and extensions on those aren't sufficient for your case?

If possible, and if the errors concern pagination itself, I'd move them downards, or if they prevent pagination from working you could consider a union instead

**Reply by xoldd** [reply]
You've heard of 200 OK graphql errors right? Returning errors as part of the data. So, that errors can be predictable and have a proper schema making consumption of api easier for clients accessing it. That's exactly what I'm doing for the query.posts field. I'm returning two nested fields `clientErrors` and `connection` in the return object for query.posts field. `connection` field returns relay-spec compliant connection object. But urql's relay plugin only seems to work when that connection object is returned directly from a query field.

**Reply by xoldd** [reply]
The console errors I posted they're happening when I'm providing `relayPagination()` to `resolvers.PostsConnectionQueryPayload` inside cacheExchange's resolver field.

**Reply by xoldd** [reply]
> I'm sorry, I don't really understand what you're trying to describe, since unfortunately all examples here you posted are of working queries and not broken ones.
> 
> However, am I correct in assuming you mean a field like `A.field(first: 10, after: $after)` which then returns a type that has a `connection` field which then leads to the real connection?
> 
> If that's the case that's not actually compliant with the Relay pagination spec: https://relay.dev/graphql/connections.htm#sec-Arguments
> 
> As stated there:
> 
> > A field that returns a Connection Type must include forward pagination arguments, backward pagination arguments, or both.
> 
> The connection type must be returned directly by the field that accepts the pagination arguments. You mentioned something like errors, so I assume you've decided that you want to cache errors and those special errors are different than GraphQLErrors and extensions on those aren't sufficient for your case?
> 
> If possible, and if the errors concern pagination itself, I'd move them downards, or if they prevent pagination from working you could consider a union instead

I've used this approach after reading this blog post :-  [productionreadygraphql/error_handling](https://productionreadygraphql.com/2020-08-01-guide-to-graphql-errors). Scroll down to the last section `Stage 6a: Error Union List + Interface`.

So, I guess it's a problem on my part. I'm going to try moving the arguments down to connection field and see if it works. Thanks for the quick reply brother.
